napari/utils/migrations.py

```diff
@@ -218,3 +245,71 @@ def __init_subclass__(cls, **kwargs):
     _OldClass.__new__.__signature__ = prealloc_signature  # type: ignore [attr-defined]
 
     return _OldClass
+
+
+class DeprecatingDict(dict[str, Any]):
```

Comment:
Sub-classing `dict` means that we don't get consistent implementations of `pop`, `get`, etc. for free as those don't necessarily call our definitions of `__setitem__` and `__getitem__`.

Instead, we'll need to sub-class `UserDict` or `MutableMapping`, which seems preferable over manually overriding `dict` methods.

This means that the return type of `_get_state` should be changed from `dict[str, Any]` to `MutableMapping[str, Any]`. That seems fine to me, but wanted to confirm.

Reply:
I'd personally prefer `MutableMapping[str, Any]` since it has a smaller API, but that could break people that have been relying on some `dict` specific methods (e.g. `copy`)